### PR TITLE
bcat & badv array merge option in stored request processing

### PR DESCRIPTION
**Pull request by postindustria-code**
### Issue

Addresses [Issue #4638](https://github.com/prebid/prebid-server/issues/4638)

### Overview

Added account-level configuration to control how array fields are merged when combining stored requests with incoming requests.
                                                                                                                                                                                                                                        
### Configuration

New field in account configuration:
```
account_defaults:
  stored_request: 
    array_merge: "concat"
```
                                                                                                                                                                                                                   
Modes:                                                                                                                                                                                                                                            
  - replace (default): Arrays are replaced following RFC 7386 JSON Merge Patch semantics                                                             
  - concat: Arrays are concatenated for specific fields (bcat, badv)
                                          
### Behavior
When `array_merge: concat` is enabled, array fields are concatenated during stored request merging instead of being replaced. Deduplication is not performed.

This applies to:
  - Stored bid request merge
  - Default stored request merge

Edge cases preserved:
  - Empty arrays in incoming request still clear stored request arrays (RFC 7386 semantics)
  - Non-array fields always follow RFC 7386 merge rules
  - Single-sided arrays (only in stored or only in incoming) are preserved as-is

Example:
```
// stored request
{
    "bcat": ["IAB25", "IAB26"],
    "badv": ["competitor.com"],
    "site": {"name": "StoredApp"}
}

// incoming request:
{
    "bcat": ["IAB8"],
    "badv": ["spam.com"],
    "site": {"name": "MyApp"}
}

// result with array_merge: "replace" (default):
{
    "bcat": ["IAB8"],
    "badv": ["spam.com"],
    "site": {"name": "MyApp"}
}

// result with array_merge: "concat":
{ 
    "bcat": ["IAB25", "IAB26", "IAB8"],
    "badv": ["competitor.com", "spam.com"], 
    "site": {"name": "MyApp"}
}
```

### CODE REVIEW

**Review by antoxas1986 — COMMENTED** [review]
## Review summary

Thanks for the PR — the feature is reasonable and the happy-path tests are thorough. I'd suggest addressing a few things before merge; there's one critical data-integrity issue plus some correctness/robustness gaps. Details inline, high level:

- 🔴 **Critical:** `concat` mode round-trips the *entire* request through `map[string]interface{}`, corrupting large/precise numbers **anywhere** in the document (verified). `replace`/`jsonpatch.MergePatch` does not have this problem.
- 🟠 **High:** concat ordering differs between the UUID and normal merge paths; `array_merge` has no validation (typos silently fall back to `replace`); error-message parity with the replace path is lost.
- 🟡 **Medium:** hardcoded `{bcat,badv}` list (re-allocated per call); empty-array / non-array edge cases undocumented & untested.

**Recommended core fix (resolves the critical issue + perf + key-reordering at once):** keep a single `jsonpatch.MergePatch(base, patch)` as today, then splice only `bcat`/`badv` onto the result with `jsonparser.Set` (the same primitive already used for `imp` further down), capturing the base arrays *before* the merge. This avoids decoding the whole document.

**Test gaps:** the `expectError` field is never exercised; no large-number regression test; no config-validation test; the ordering tests currently encode the inconsistency rather than guard against it.

File: endpoints/openrtb2/auction.go [hunk 1]
```diff
@@ -1689,7 +1689,69 @@ func (deps *endpointDeps) getStoredRequests(ctx context.Context, requestJson []b
...
+func mergeWithArrayConcat(base, patch []byte, arrayFields []string) ([]byte, error) {
+	if len(base) == 0 {
+		return patch, nil
+	}
+	if len(patch) == 0 {
+		return base, nil
+	}
+
+	// Parse both JSON documents
+	var baseObj, patchObj map[string]interface{}
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🔴 **Critical — numeric precision regression.** Decoding the full `base`/`patch` into `map[string]interface{}` forces every JSON number to `float64`; the re-marshal below then corrupts any integer > 2^53 or high-precision float — across the **whole** request, not just `bcat`/`badv` (`tmax`, floor prices, large IDs, numeric `ext`, …).

This is a regression vs. the existing path: `jsonpatch.MergePatch` keeps untouched values as raw bytes and never round-trips scalars through `float64`. Reproduced against this repo's `jsonutil`:

```
REPLACE (today): {"price":1234567890123456789,"tmax":12345678901234567}   <- exact
CONCAT  (PR):    {"price":1234567890123456800,"tmax":12345678901234568}   <- corrupted
```

**Suggested fix** — run a single `MergePatch`, then splice only the target arrays onto the result via `jsonparser.Set` (the primitive already used for `imp`), reading the arrays back from the original `base`/`patch`. This keeps every number byte-for-byte and removes the redundant parses. I verified this against this package (numbers stay exact; `["IAB1","IAB2"] + ["IAB3"]` → `["IAB1","IAB2","IAB3"]`; `site` still RFC-7386 merged):

```go
var arrayConcatFields = []string{"bcat", "badv"}

// mergeWithArrayConcat merges patch into base with RFC 7386 JSON Merge Patch
// semantics, except that for the given fields the base and patch arrays are
// concatenated (base elements first) instead of replaced. All other fields —
// and their number formatting — are preserved by MergePatch. If a field is
// absent, not an array, or empty on either side, replace semantics apply.
func mergeWithArrayConcat(base, patch []byte, arrayFields []string) ([]byte, error) {
    if len(base) == 0 {
        return patch, nil
    }
    if len(patch) == 0 {
        return base, nil
    }

    merged, err := jsonpatch.MergePatch(base, patch)
    if err != nil {
        return nil, err
    }

    for _, field := range arrayFields {
        baseArr, baseType, _, baseErr := jsonparser.Get(base, field)
        if baseErr != nil || baseType != jsonparser.Array {
            continue
        }
        patchArr, patchType, _, patchErr := jsonparser.Get(patch, field)
        if patchErr != nil || patchType != jsonparser.Array {
            continue
        }

        var baseElems, patchElems []json.RawMessage
        if err := jsonutil.Unmarshal(baseArr, &baseElems); err != nil {
            return nil, err
        }
        if err := jsonutil.Unmarshal(patchArr, &patchElems); err != nil {
            return nil, err
        }
        // Empty array on either side keeps RFC 7386 replace semantics
        // (e.g. an explicit [] in the patch clears the field).
        if len(baseElems) == 0 || len(patchElems) == 0 {
            continue
        }

        combined, err := jsonutil.Marshal(append(baseElems, patchElems...))
        if err != nil {
            return nil, err
        }
        if merged, err = jsonparser.Set(merged, combined, field); err != nil {
            return nil, err
        }
    }

    return merged, nil
}
```

This single rewrite also resolves the perf comment (one parse, no extra marshals), the empty/non-array edge cases, and keeps error behaviour identical to the replace path (its only error source is the same `MergePatch`). (Minor: prefer `any` over `interface{}` per this file — moot once the maps are gone.)

_Correction to my earlier note: key reordering is **not** a regression — `MergePatch` already alphabetizes keys in the replace path too, so this is unchanged either way._

File: endpoints/openrtb2/auction.go [hunk 2]
```diff
@@ -1689,7 +1689,69 @@ func (deps *endpointDeps) getStoredRequests(ctx context.Context, requestJson []b
...
+	baseBytes, err := jsonutil.Marshal(baseObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal base object: %w", err)
+	}
+	patchBytes, err := jsonutil.Marshal(patchObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal patch object: %w", err)
+	}
+
+	return jsonpatch.MergePatch(baseBytes, patchBytes)
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🟠 **Performance (hot path).** Even setting the precision issue aside, concat mode parses the document 3× (unmarshal `base` + `patch` above, then `MergePatch` parses both again here) and marshals 2×, per merge call — up to 3 calls per request.

**Suggested fix:** the splice-only rewrite in the critical comment (single `MergePatch` + `jsonparser.Set` for just `bcat`/`badv`) collapses this to one parse and removes both extra marshals — no separate change needed here once that lands.

File: endpoints/openrtb2/auction.go [hunk 3]
```diff
@@ -1689,7 +1689,69 @@ func (deps *endpointDeps) getStoredRequests(ctx context.Context, requestJson []b
 	return storedBidRequestId, hasStoredBidRequest, storedRequests, storedImps, errs
 }
 
-func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []ImpExtPrebidData, storedRequests map[string]json.RawMessage, storedImps map[string]json.RawMessage, storedBidRequestId string, hasStoredBidRequest bool) ([]byte, map[string]exchange.ImpExtInfo, []error) {
+// mergeWithArrayConcat performs JSON merge with array concatenation for specific fields.
+// For the specified arrayFields (e.g., "bcat", "badv"), arrays are concatenated
+// All other fields follow RFC 7386 JSON Merge Patch semantics.
+// Empty arrays inside the patch request will clear the base request arrays (preserving RFC 7386 semantics).
+func mergeWithArrayConcat(base, patch []byte, arrayFields []string) ([]byte, error) {
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🟠 **High — error-message parity with the replace path is lost.** In replace mode, `MergePatch` failures are routed through `storedRequestErrorChecker` / `getJsonSyntaxError`, producing client-facing messages like `"Invalid JSON in Incoming Request: …"`. This function instead returns raw `fmt.Errorf("failed to unmarshal base request: %w", …)`, so the *same* malformed request yields a different, un-classified, lowercase error depending on the mode.

**Suggested fix:** the splice rewrite in the critical comment removes the whole-document unmarshal entirely — its only error source is the same `jsonpatch.MergePatch` call as the replace path, so the existing `storedRequestErrorChecker` handling at the call sites applies unchanged. No bespoke error strings needed.

File: endpoints/openrtb2/auction.go [hunk 4]
```diff
@@ -1689,7 +1689,69 @@ func (deps *endpointDeps) getStoredRequests(ctx context.Context, requestJson []b
...
+		baseArr, ok := baseVal.([]interface{})
+		if !ok {
+			continue
+		}
+		patchArr, ok := patchVal.([]interface{})
+		if !ok {
+			continue
+		}
+
+		if len(patchArr) > 0 && len(baseArr) > 0 {
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🟡 **Medium — edge cases undocumented & untested.**
- The `len(patchArr) > 0 && len(baseArr) > 0` guard means an incoming `"bcat": []` falls through to replace and wipes stored values. The end result is arguably right, but please document and test all four combinations (empty/empty, empty/full, full/empty, full/full).
- If `bcat`/`badv` is present but not an array (malformed stored request, or `null`), the type assertions above fall through to replace silently — probably the safe choice, but make it deliberate.

**Suggested fix:** the splice rewrite in the critical comment handles both explicitly — `baseType != jsonparser.Array` covers non-array/`null`, and `if len(baseElems) == 0 || len(patchElems) == 0 { continue }` covers empty arrays — so the behaviour is intentional and self-documenting. Please still add the four-combination tests plus a non-array case.

File: endpoints/openrtb2/auction.go [hunk 5]
```diff
@@ -1698,6 +1760,14 @@ func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []Im
 	// Apply the Stored BidRequest, if it exists
 	resolvedRequest := requestJson
 
+	useConcatMode := account != nil && account.StoredRequest.ArrayMerge == config.ArrayMergeConcat
+	merge := func(base, patch []byte) ([]byte, error) {
+		if useConcatMode {
+			return mergeWithArrayConcat(base, patch, []string{"bcat", "badv"})
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🟡 **Medium — hardcoded field list, re-allocated per call.** `[]string{"bcat", "badv"}` is allocated on every merge invocation; hoist it to the package-level `var arrayConcatFields` introduced in the critical-comment rewrite (this committable suggestion depends on that var existing):

```suggestion
			return mergeWithArrayConcat(base, patch, arrayConcatFields)
```

Also, the generic config name `array_merge` implies broad applicability, but only these two fields are affected — sibling arrays (`wseat`/`bseat`, `bapp`, `wlang`, `cur`) have the same accumulation semantics. Consider documenting the limited scope or making the field set configurable.

File: endpoints/openrtb2/auction.go [hunk 6]
```diff
@@ -1713,13 +1783,14 @@ func (deps *endpointDeps) processStoredRequests(requestJson []byte, impInfo []Im
 				errL := storedRequestErrorChecker(requestJson, storedRequests, storedBidRequestId)
 				return nil, nil, errL
 			}
-			resolvedRequest, err = jsonpatch.MergePatch(requestJson, uuidPatch)
+
+			resolvedRequest, err = merge(requestJson, uuidPatch)
```
Comment (antoxas1986) [inline endpoints/openrtb2/auction.go]: 🟠 **High — concat ordering is inverted between paths.**
- Non-UUID path (`merge(storedRequests[...], requestJson)`, line 1793): base=stored, patch=incoming → concat order **[stored, incoming]**.
- This UUID path (`merge(requestJson, uuidPatch)`): base=incoming, patch=stored+uuid → concat order **[incoming, stored]**.

So `bcat`/`badv` element order depends on whether UUID generation triggers (app request + `GenerateRequestID`, or `id == "{{UUID}}"`). The added tests bake in the divergence (`["IAB1","IAB3"]` vs `["IAB3","IAB1","IAB2"]`). How much this matters depends on whether anything downstream treats these as ordered (for most bidders they're set-like) — but it's a latent trap.

⚠️ **Careful with the fix:** the two paths also have *different scalar precedence today* — the non-UUID path lets the incoming request win, the UUID path lets the stored request win (see the existing `"UUID generation with replace mode - stored request wins"` test). That precedence is pre-existing and intentional, so **don't** simply swap the UUID-path args to `merge(stored, incoming)` — that would flip precedence and break that test.

**Suggested fix:** normalize only the concat *order*, not precedence. Either (a) document the difference and have each test assert its own order deliberately, or (b) if you want one canonical order, concatenate in a fixed `[stored, incoming]` order independent of which side is the merge base — e.g. give the concat helper the stored and incoming arrays explicitly rather than inferring order from base/patch. Pick one and lock it in with tests.

File: config/account.go [hunk 7]
```diff
@@ -395,3 +396,18 @@ func (ip *IPv4) Validate(errs []error) []error {
 	}
 	return errs
 }
+
+// AccountStoredRequest represents account-specific stored request configuration
+type AccountStoredRequest struct {
+	ArrayMerge ArrayMergeMode `mapstructure:"array_merge" json:"array_merge"`
```
Comment (antoxas1986) [inline config/account.go:404]: 🟠 **High — no validation; typos silently fall back to `replace`.** `ArrayMerge` accepts any string, and the call site only checks `== ArrayMergeConcat`, so `"concatt"`, `"Concat"`, etc. silently behave as `replace` with no error.

**Suggested fix** — add a `validate` method (plus doc comments per the project convention) and wire it into `(*Configuration).validate` next to `PriceFloors` (config.go:176). `fmt` is already imported in this file:

```go
// AccountStoredRequest configures how stored requests are merged for an account.
type AccountStoredRequest struct {
    ArrayMerge ArrayMergeMode `mapstructure:"array_merge" json:"array_merge"`
}

func (sr *AccountStoredRequest) validate(errs []error) []error {
    switch sr.ArrayMerge {
    case "", ArrayMergeReplace, ArrayMergeConcat:
    default:
        errs = append(errs, fmt.Errorf(`account_defaults.stored_request.array_merge: invalid value %q (allowed: "replace", "concat")`, sr.ArrayMerge))
    }
    return errs
}
```

```go
// in (*Configuration).validate, alongside the other AccountDefaults checks:
errs = cfg.AccountDefaults.StoredRequest.validate(errs)
```

Confirm per-account stored JSON is validated through the same path (not just host `account_defaults`), and add a config test mirroring `TestAccountPriceFloorsValidate`.

**Comment by postindustria-code** [comment]
@antoxas1986 

Thanks for the thorough review — all points addressed and pushed.
>
> **Numeric precision (critical) — fixed.** Reworked `mergeWithArrayConcat` to the splice approach you
> suggested: a single `jsonpatch.MergePatch(base, patch)`, then for each of `bcat`/`badv` read the source
> arrays with `jsonparser` and splice the concatenation back onto the merged result via `jsonparser.Set`
> (the primitive the `imp` handling already uses). No more full-document `map[string]interface{}` round-trip,
> so every untouched number stays byte-exact. Locked by a new `TestMergeWithArrayConcatPreservesNumbers`
> asserting a large `tmax`/`ext.price` survive verbatim.
>
> **Concat ordering — fixed, precedence preserved.** Good catch on the precedence trap. I left each path's
> `base`/`patch` exactly as-is (so the "stored wins" UUID behavior is unchanged) and added an explicit
> `incomingIsBase` flag that only controls concat order. Canonical order is now `[incoming, stored]` on all
> three paths (UUID, non-UUID, default-request) — the incoming request's entries first, stored/host
> additions appended. That matches the framing in #4638 ("the publisher is already sending a list … the host
> wants to add to these") and PBS's "defer to the HTTP request" convention. The ordering tests now assert the
> same order on both paths (they previously baked in the divergence).
>
> **Error-message parity — fixed.** The splice rewrite drops the bespoke unmarshal error strings; the only
> error source is the same `MergePatch` as the replace path, so the existing `storedRequestErrorChecker`
> wrapping at the call sites applies unchanged.
>
> **Config validation — fixed.** Added `(*AccountStoredRequest).validate`, wired into `Configuration.validate`
> next to `PriceFloors`; unknown `array_merge` values are now rejected. This validates `account_defaults`,
> mirroring the `PriceFloors` precedent — per-account fetched configs still fall back to replace on an unknown
> value, and I documented that scope on the method.
>
> **Hardcoded field list — fixed.** Hoisted to a package-level `var arrayConcatFields`, and documented the
> bcat/badv-only scope on `ArrayMergeMode` per your note about the generic `array_merge` name.
>
> **Edge cases + tests.** Empty-array (clears, replace semantics) and non-array/`null` (replace) are handled
> explicitly and documented in the function comment, with tests for all four empty/non-array combinations.
> Also added the large-number regression, a config-validation test, exercised `expectError`, and rewrote the
> ordering tests to guard the canonical order on both paths.

**Review by antoxas1986 — APPROVED** [review]
Re-reviewed after the 2026-07-23 push. All points from my previous review are addressed and verified against the actual code (not just the summary above) — thank you for the thorough rework:

- **Numeric precision** — confirmed fixed. The splice approach (single `MergePatch` + `jsonparser.Get`/`Set` for `bcat`/`badv` only) never round-trips the document through `map[string]interface{}`. `TestMergeWithArrayConcatPreservesNumbers` uses values `> 2^53`, so it genuinely catches the old failure mode.
- **Performance** — resolved as a side effect of the above (one parse, no extra marshals).
- **Error parity** — confirmed; `MergePatch` is the only error source again, so the existing `storedRequestErrorChecker` wrapping applies unchanged.
- **Edge cases** — documented and covered by 12 sub-tests across the empty/non-array combinations.
- **Hardcoded field list** — hoisted to package-level `arrayConcatFields`, scope documented.
- **Concat ordering vs. precedence** — this was the trickiest one. Traced all three call sites (UUID, non-UUID, default-request) by hand: `incomingIsBase` only affects array-splice order, the actual `MergePatch(base, patch)` arguments are unchanged from before at every site, so each path's pre-existing scalar precedence is untouched while concat order is now canonical (incoming-first) everywhere — including correctly cascading through the default-request path. Nicely done.
- **Config validation** — `AccountStoredRequest.validate` mirrors the `PriceFloors` pattern correctly and is wired in at the right place, with a matching test.

`go build`/`go vet`/`gofmt`/`golangci-lint` clean, full suite passes.

Two optional, non-blocking nits — feel free to take or leave:
1. In `TestProcessStoredRequests_DefaultRequestArrayMerge`, the sub-test named "no default stored request - incoming unchanged" actually leaves `deps.defaultRequest = true` and exercises the empty-`defReqJSON` short-circuit instead — correct behavior, just a slightly misleading name.
2. No test currently drives a concat-mode malformed-JSON error all the way through `processStoredRequests` (only `mergeWithArrayConcat`'s own error return is unit-tested) — low risk since the error wrapping is untouched, but would round out coverage.
